lib/bolt/transport/api.rb

```diff
@@ -2,26 +2,213 @@
+          end
+        end
+      end
+
+      def run_task(targets, task, arguments, options = {})
+        run_task_job(targets, task, arguments, options)
+      end
+
+      def run_command(targets, command, options = {})
+        params = { 'command' => command }
+        run_task_job(targets, BOLT_COMMAND_TASK, params, options)
+      end
+
+      def run_script(targets, script, arguments, options = {})
+        content = File.open(script, &:read)
+        content = Base64.encode64(content)
+        params = {
+          'content' => content,
+          'arguments' => arguments
+        }
+        callback ||= proc {}
+        results = run_task_job(targets, BOLT_SCRIPT_TASK, params, options, &callback)
+        results.map! { |result| unwrap_bolt_result(result.target, result) }
+        results.each do |result|
+          callback.call(type: :node_result, result: result)
+        end
+      end
+
+      def pack(directory)
+        start_time = Time.now
+        io = StringIO.new
+        output = Minitar::Output.new(Zlib::GzipWriter.new(io))
+        Find.find(directory) do |file|
+          next unless File.file?(file)
+
+          tar_path = Pathname.new(file).relative_path_from(Pathname.new(directory))
+          @logger.debug("Packing #{file} to #{tar_path}")
+          stat = File.stat(file)
+          content = File.binread(file)
+          output.tar.add_file_simple(
+            tar_path.to_s,
+            data: content,
+            size: content.size,
+            mode: stat.mode & 0o777,
+            mtime: stat.mtime
+          )
+        end
+
+        duration = Time.now - start_time
+        @logger.debug("Packed upload in #{duration * 1000} ms")
+
+        output.close
+        io.string
+      ensure
+        # Closes both tar and sgz.
+        output&.close
+      end
+
+      def file_upload(targets, source, destination, options = {})
+        stat = File.stat(source)
+        content = if stat.directory?
+                    pack(source)
+                  else
+                    File.open(source, &:read)
+                  end
+        content = Base64.encode64(content)
+        mode = File.stat(source).mode
+        params = {
+          'path' => destination,
+          'content' => content,
+          'mode' => mode,
+          'directory' => stat.directory?
+        }
+        results = run_task_job(targets, BOLT_UPLOAD_TASK, params, options)
+        results.map! do |result|
+          if result.error_hash
+            result
+          else
+            Bolt::Result.for_upload(result.target, source, destination)
+          end
+        end
+      end
+
+      # Unwraps any Sensitive data in an arguments Hash, so the plain-text is passed
+      # to the Task/Script.
+      #
+      # This works on deeply nested data structures composed of Hashes, Arrays, and
+      # and plain-old data types (int, string, etc).
+      def unwrap_sensitive_args(arguments)
```

Comment:
It would be good to move this to a location where it can be shared. Maybe a class method on `Bolt::PAL`, since it does explicitly Puppet-y things.

Reply:
I'd kind of like to ticket this and do it later in the interest of time, but if you're ok with waiting until Monday or doing it yourself I'm not opposed to having it be part of this PR

Reply:
Doing it later is fine.

Reply:
BOLT-1102 :+1: